### PR TITLE
Fix credentials for Tower API V2

### LINKED ISSUE #31655: tower_* modules not updated for Tower 3.2/ tower_cli 3.2
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
tower_credential
tower_*

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/notting/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

##### CONFIGURATION
```
ANSIBLE_SSH_PIPELINING(/etc/ansible/ansible.cfg) = True
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
PERSISTENT_CONNECT_TIMEOUT(/etc/ansible/ansible.cfg) = 30
```

##### OS / ENVIRONMENT

Fedora 26

##### SUMMARY

With tower_cli 3.2.0 installed, assorted Tower modules fail:

```
[notting@nostromo: ~]$ ansible -m tower_credential localhost -a "kind=ssh username=wow password=wow name=frobozznicate organization=TGP"
[WARNING]: Could not match supplied host pattern, ignoring: all

[WARNING]: provided hosts list is empty, only localhost is available

An exception occurred during task execution. To see the full traceback, use -vvv. The error was: tower_cli.exceptions.BadRequest: Missing required fields: credential-type
localhost | FAILED! => {
   "changed": false,
   "failed": true,
   "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_imDa8H/ansible_module_tower_credential.py\", line 292, in <module>\n    main()\n  File \"/tmp/ansible_imDa8H/ansible_module_tower_credential.py\", line 277, in main\n    result = credential.modify(**params)\n  File \"/usr/lib/python2.7/site-packages/tower_cli/models/base.py\", line 659, in modify\n    return self.write(pk, create_on_missing=create_on_missing, force_on_exists=True, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/tower_cli/models/base.py\", line 357, in write\n    raise exc.BadRequest('Missing required fields: %s' % ', '.join(missing_fields).replace('_', '-'))\ntower_cli.exceptions.BadRequest: Missing required fields: credential-type\n",
   "module_stdout": "",
   "msg": "MODULE FAILURE",
   "rc": 0
}
```

##### STEPS TO REPRODUCE

see above

##### EXPECTED RESULTS

works

##### ACTUAL RESULTS

attempts to use /api/v1 parameters from 3.1 and earlier with tower-cli 3.2's /api/v2 usage. Things go kablooie.

### PR DESCRIPTION
**Pull request by tbouvet**
##### SUMMARY
When passing **params directly to credential.modify(), you get a rejection
 by the API due to fields that it doesn't accept (such as 'state'). Fix for API V2

Fixes #31655

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ansible_tower

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

### CODE REVIEW

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 1]
```diff
@@ -213,22 +218,57 @@ def main():
...
+                filename = module.params.get('ssh_key_data')
                 if not os.path.exists(filename):
                     module.fail_json(msg='file not found: %s' % filename)
                 if os.path.isdir(filename):
                     module.fail_json(msg='attempted to read contents of directory: %s' % filename)
                 with open(filename, 'rb') as f:
-                    params['ssh_key_data'] = f.read()
+                    params['inputs']['ssh_key_data'] = f.read()
+
+            _add_input_(module, 'authorize', params['inputs'])
```
Comment (ryanpetrello) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py:257]: My only feedback is that this is a _bit_ repetitive (would probably be better as a loop over a list of these keys) though I don't think I'd let it hold up a merge.

**Comment by ryanpetrello** [comment]
This change _does_ hard-code v1 credential fields - the new Credential Type model (where you specify the type and a list of inputs) _is_ more flexible, and this change doesn't support the flexibility it provides in custom credential types.

That said, I think given that _right now_ this is all just totally broken, I'm inclined to 👍 and merge this to get things moving on https://github.com/ansible/ansible/issues/31655

**Comment by matburt** [comment]
shipit

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 2]
```diff
@@ -213,22 +218,57 @@ def main():
...
+            credential_type_res = tower_cli.get_resource('credential_type')
+            credential_type = credential_type_res.get(kind=module.params.get('kind'))
+            params['credential_type'] = credential_type['id']
+
+            if module.params.get('description'):
+                params['description'] = module.params.get('description')
+
+            if module.params.get('user'):
+                user_res = tower_cli.get_resource('user')
+                user = user_res.get(name=module.params.get('user'))
```
Comment (AlanCoding) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py]: Wouldn't you need to filter by `username`, not `name`?

File: lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py [hunk 3]
```diff
@@ -213,22 +218,57 @@ def main():
...
+                params['user'] = user['id']
+
+            if module.params.get('team'):
+                team_res = tower_cli.get_resource('team')
+                team = team_res.get(name=module.params.get('team'))
```
Comment (AlanCoding) [inline lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py:244]: If both team and organization is specified, then I would think we would filter the team by both, and remove the organization from params. This is for uniqueness (not sure if we've really dealt with that in other cases). Note: team is only valid on creation.

**Comment by AlanCoding** [comment]
I think this should fix the `username` bug, and that should be sufficient to merge.

```
>>> import tower_cli
>>> user_res = tower_cli.get_resource('user')
>>> user_res.get(name='alan')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "tower_cli/models/base.py", line 484, in get
    response = self.read(pk=pk, fail_on_no_results=True, fail_on_multiple_results=True, **kwargs)
  File "tower_cli/models/base.py", line 286, in read
    r = client.get(url, params=kwargs)
  File "/Users/alancoding/.virtualenvs/cli/lib/python2.7/site-packages/requests-2.18.4-py2.7.egg/requests/sessions.py", line 521, in get
    return self.request('GET', url, **kwargs)
  File "tower_cli/api.py", line 236, in request
    kwargs.get('data', None), r.content.decode('utf8'))
tower_cli.exceptions.BadRequest: The Tower server claims it was sent a bad request.

GET http://localhost:8013/api/v2/users/
Params: {'name': 'alan'}
Data: None

Response: {"detail":"User has no field named 'name'"}
```

**Comment by ryanpetrello** [comment]
@AlanCoding @tbouvet I've opened a new PR with various fixes discussed here: https://github.com/ansible/ansible/pull/36587

**Comment by tbouvet** [comment]
@AlanCoding `username `bug is fixed.

**Comment by AlanCoding** [comment]
The PR 36587 still has your original commit in it, and includes a number of important fixes, so I think we're best going with that one. (it should be ready to merge already)

**Comment by samdoran** [comment]
Closing as this is fixed in PR #36587

**Pull request closed without merging**